File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const visited = new Array(words.length).fill(false);
    const wordLen = target.length;
    
    const queue = [[begin, 0]];
    
    while (queue.length>0) {
        const [word, cnt] = queue.shift();
        
        if (word===target) return cnt;
        
        for (let i=0; i<words.length; i++) {
            if (visited[i]) continue;
            
            let diff = 0;
            for (let j=0; j<wordLen; j++) {
                if (words[i][j] !== word[j]) {
                    diff++;
                    if (diff>1) break;
                }
            }
            
            if (diff===1) {
                queue.push([words[i], cnt+1]);
                visited[i] = true;
            }
        }
    }
    
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    q = deque()
    visit = [0 for _ in range(len(words))]
    q.append((begin, 0, visit))
    answer = 0
    
    while q:
        word, cnt, v = q.popleft()
        
        if word==target:
            return cnt
        
        for i in range(len(words)):
            if v[i]==0:
                diff = 0
                for j in range(len(word)):
                    if words[i][j]!=word[j]:
                        diff+=1
                if diff==1:
                    q.append((words[i], cnt+1, v[:i]+[1]+v[i+1:]))
    return 0